### PR TITLE
Create Sonicwall log parser plugin

### PR DESCRIPTION
**Pull request by ericwholt**
This parser will parse Sonicwall logs. This plugin uses key_value parser which requires stanza 1.2.7. Used https://www.sonicwall.com/techdocs/pdf/sonicos-6-5-4-log-events-reference-guide.pdf as reference for fields. It handles two log types. One that starts with a syslog priority format `<134>` and another that starts with `Apr 1 10:45:16 10.1.5.1` these were the only example logs I had available at the time of creating this plugin.

Example Output
``` JSON
{
  "timestamp": "2004-04-01T10:39:35-05:00",
  "severity": 40,
  "severity_text": "5",
  "labels": {
    "file_name": "logs.log",
    "file_name_resolved": "logs.log",
    "file_path": "/Users/ericholt/bindplane/bplogagent-dev/sonicwall-logs/logs.log",
    "file_path_resolved": "/Users/ericholt/bindplane/bplogagent-dev/sonicwall-logs/logs.log",
    "log_type": "sonicwall",
    "plugin_id": "sonicwall"
  },
  "record": {
    "c": "64",
    "dst": "67.32.44.2:445:LAN",
    "fw": "67.32.44.2",
    "host": "10.1.5.1",
    "id": "firewall",
    "m": "36",
    "msg": "TCP connection dropped",
    "n": "2686",
    "pri": "5",
    "proto": "tcp",
    "sn": "00301E0526B1",
    "src": "67.101.200.27:4507:WAN"
  }
}
```

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.0.80] Unreleased
```
Comment (jsirianni) [inline CHANGELOG.md:8]: 0.0.81

File: plugins/sonicwall.yaml [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+  - id: beginning_timestamp_parser
+    if: '$record.log matches "^\\w{3}\\s+\\d{1,2}\\s+\\d{2}:\\d{2}:\\d{2}\\s+"'
+    type: regex_parser
+    parse_from: $record.log
+    regex: '^(?P<timestamp>\w{3}\s+\d{1,2}\s+\d{2}:\d{2}:\d{2})\s+(?P<host>[:a-fA-F\d\.]+)\s+(?P<message>[\d\D]*)'
+    timestamp:
+      parse_from: $record.timestamp
+      layout_type: gotime
+      layout: 'Jan _2 15:04:05'
```
Comment (jsirianni) [inline plugins/sonicwall.yaml:110]: We should support a timezone when timezones do not exist in the log format

File: plugins/sonicwall.yaml [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      debug: [7,15,23,31,39,47,55,63,71,79,87,95,103,111,119,127,135,143,151,159,167,175,183,191]
+    output: key_value_parser
+
+  - type: key_value_parser
+    if: '$record.message != nil'
+    parse_from: message
+    timestamp:
+      parse_from: time
+      layout_type: gotime
+      layout: '2006-01-02 15:04:05'
```
Comment (jsirianni) [inline plugins/sonicwall.yaml:136]: We should support a timezone when timezones do not exist in the log format

**Review by jsirianni — CHANGES_REQUESTED** [review]
Couple additional things:
1. File input likely is not approbate. I know we were not sure what the transport medium would be, so this is fine for now but I suspect we will switch to udp / tcp before merging.
2. Tests are failing, update go.mod to use stanza 1.2.7, which will give you the k/v parser.

File: CHANGELOG.md (hunk 1 above)
Comment (jsirianni) [reply CHANGELOG.md:8]: 0.0.83 now, we released twice

**Comment by jsirianni** [comment]
Lets rename msg to message

File: plugins/sonicwall.yaml (hunk 3 above)
Comment (ericwholt) [reply plugins/sonicwall.yaml:144]: Done

File: plugins/sonicwall.yaml (hunk 2 above)
Comment (ericwholt) [reply plugins/sonicwall.yaml:117]: Done

File: CHANGELOG.md (hunk 1 above)
Comment (ericwholt) [reply CHANGELOG.md]: Merged Master and updated

**Review by jsirianni — CHANGES_REQUESTED** [review]
Please make the following changes
- file_input --> UDP
- provide sample logs

Otherwise I think it looks good

File: plugins/sonicwall.yaml [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    default: "UTC"
+  - name: start_at
+    label: Start At
+    description: Start reading file from 'beginning' or 'end'
+    type: enum
+    valid_values:
+     - beginning
+     - end
+    default: end
```
Comment (jsirianni) [inline plugins/sonicwall.yaml:28]:
```suggestion
```
not required

**Review by jsirianni — CHANGES_REQUESTED** [review]
Not sure what is happening,

you can send over udp like this:
```
echo '<134>  id=LVM_Sonicwall sn=18B1693B2C90 time="2021-09-13 18:16:13" fw=142.166.177.10 pri=6 c=262144 gcat=6 m=98 msg="Connection Opened" src=192.168.11.53:36510:X0 natSrc=142.166.177.10:51879 dst=142.250.64.68:443:X1 natDst=142.250.64.68:443 proto=tcp/https sent=60 n=53798' | nc -u localhost 5140
```

config
```yaml
pipeline:
- type: sonicwall
  listen_port: 5140
  location: UTC
- type: stdout
```

If I use just udp, it works fine and has output. With the sonicwall plugin I do not get output.

**Review by jsirianni — APPROVED** [review]

**Comment by jsirianni** [comment]
```json
{
  "timestamp": "2021-09-13T18:16:13Z",
  "severity": 30,
  "severity_text": "134",
  "labels": {
    "log_type": "sonicwall",
    "net.host.ip": "::",
    "net.host.port": "5555",
    "net.peer.ip": "::1",
    "net.peer.port": "33248",
    "net.transport": "IP.UDP",
    "plugin_id": "sonicwall"
  },
  "record": {
    "c": "262144",
    "dst": "142.250.64.68:443:X1",
    "fw": "142.166.177.10",
    "gcat": "6",
    "id": "LVM_Sonicwall",
    "m": "98",
    "message": "Connection Opened",
    "n": "53798",
    "natDst": "142.250.64.68:443",
    "natSrc": "142.166.177.10:51879",
    "pri": "6",
    "priority": "134",
    "proto": "tcp/https",
    "sent": "60",
    "sn": "18B1693B2C90",
    "src": "192.168.11.53:36510:X0"
  }
}

```

**Pull request merged**